### ISSUE TITLE
Cursor not showing when using MobileVNC server

### ISSUE THREAD

**Issue by dvader** [novnc/noVNC#1031]
Is there a way to force the cursor on or to get noVNC to properly recognize my server as a server that has a cursor?

The VNC cursor is not showing up in noVNC when I connect to my touch screen/tablet (HMI). The HMI is running on Windows CE and it is running smartlab.at MobileVNC server (http://www.smartlab.at/MobileVNC/). I am using noVNC 1.0.0. I tried Firefox and Chrome with the same results. The cursor shows up on the HMI but not in the browser. I tried RealVNC and the cursor shows up as a small black square in the client and shows up on the HMI.
Here is my canvas information:
`<canvas style="margin: auto; outline: medium none currentcolor; flex-shrink: 0; width: 800px; height: 480px; cursor: none;" width="800" height="480" tabindex="-1"></canvas>` 
(see `cursor:none`)

I was able to get the cursor to show in the browser by "hacking" the `disableLocalCursor` to be `default` instead of `none`:
``` javascript
    disableLocalCursor: function () {
        this._target.style.cursor = "default";
    },
```

**Comment by CendioOssman** [comment]
To clarify, the described behaviour indicates that the server has no support for cursors at all. RealVNC (and derived clients) have a setting to show a small local cursor whenever the server fails to provide one. The problem with this approach is that you cannot tell the difference between a server side cursor and a missing one, so you are forced to ask the user.

**Comment by CendioOssman** [comment]
Have you tried asking MobileVNC if there is some way to get their server to handle cursors? That would be the ideal and would avoid options you have to fiddle with.

**Comment by dvader** [comment]
I'm confused. When I force the cursor on in noVNC, the cursor does display in both noVNC and the HMI. Are you saying that the VNC server is returning a "supported features" that doesn't include cursor? In the end, woudn't it be better to show the cursor based on what the **client** handles instead of what the server handles? I am running noVNC on a desktop computer with no touch support so having no cursor on the client does not work.
I'm sorry but this is a purchased/fixed platform, I don't control the server configuration or even which server to use.

**Comment by DirectXMan12** [comment]
So, basically, the way things work in noVNC is that, on start, we disable the local cursor, because the "assumed default" in the protocol is that the server displays a cursor on it's end.  Then, during negotiation, (assuming the browser supports cursor URIs, and the local device isn't a touch device), we indicate that we support local cursors by sending the cursor pseudo-encoding.  If the server also supports cursor pseudo-encoding, it will send a cursor message to us containing the image we should use for a cursor, at which point we turn local cursor back on, because we assume the server has stopped displaying a cursor on its end.

This falls apart a bit if the server breaks the assumption that it *has* a cursor in the first place.  If it doesn't have a cursor, it can't display one locally, but also will never send us a cursor image, and we basically have no way to know that.  I believe @CendioOssman is conjecturing that that's happening.

**Comment by DirectXMan12** [comment]
FWIW, I'm curious what RealVNC is doing here.

**Comment by dvader** [comment]
RealVNC was just there to test the behavior of another client with the same server.

**Comment by dvader** [comment]
> the "assumed default" in the protocol is that the server displays a cursor on it's end
> This falls apart a bit if the server breaks the assumption that it has a cursor in the first place. If it doesn't have a cursor, it can't display one locally, but also will never send us a cursor image, and we basically have no way to know that.

Fair enough. So I'll go back to my initial question, is there any way to indicate to noVNC that the server doesn't display a local cursor?

**Comment by DirectXMan12** [comment]
> RealVNC was just there to test the behavior of another client with the same server.

Oh, I meant, "I'm curious how RealVNC handles this case" ;-)

> So I'll go back to my initial question, is there any way to indicate to noVNC that the server doesn't display a local cursor?

Unfortunately, IIRC, there's not a good way right now.  Just to confirm though, the table does have a cursor?  It's weird that the server isn't sending it.

**Comment by dvader** [comment]
> Oh, I meant, "I'm curious how RealVNC handles this case" ;-)

Too funny. I don't know. If I had to guess, they always display the little black dot unless the server expressly tell them that it handles the cursor display. So worst case is that the server shows the cursor but does not signal to the client that it does and the client show the back dot on top. I believe that they take the opposite approach: always assume that the server doesn't display a cursor and display the local cursor and don't display the local cursor only when the server explicitly tells you they display it on their end. 
> Unfortunately, IIRC, there's not a good way right now. Just to confirm though, the tablet does have a cursor? It's weird that the server isn't sending it.

The tablet does not have a cursor unless VNC is connected. It is a pure touch interface but it shows a cursor so the local user can see what the remote VNC is doing. But you are correct though, there is a cursor when the VNC client is connected and the VNC server isn't sending it.
If it is the only, way, I'll keep the hack to always display the local cursor.

**Comment by CendioOssman** [comment]
> FWIW, I'm curious what RealVNC is doing here.

Their client has an option to show a 3x3 pixel black square whenever the server fails to send a cursor (or sends an empty cursor). The same option appears in many RealVNC derived clients, such as TigerVNC. It is disabled by default though.

**Comment by dvader** [comment]
> Their client has an option to show a 3x3 pixel black square whenever the server fails to send a cursor (or sends an empty cursor). The same option appears in many RealVNC derived clients, such as TigerVNC. It is disabled by default though.

Thanks, that's very good info. On my version of RealVNC, this is set to ON by default.
![image](https://user-images.githubusercontent.com/2127127/36802858-a558443c-1c7b-11e8-8531-f6ec7f3c812e.png)
Maybe noVNC could offer a similar option?

**Comment by patrakov** [comment]
I think that there is another important use case here, besides the strange server.

Some proprietary Windows software that I would rather not name here, during its normal operation, hides the Windows cursor in its window and draws the equivalent by itself. noVNC shows this fake cursor just fine (because it is a part of the remote image), but, if the server is very far (in China), then the lag creates usability issues.

So, I would like to see a dot if there is no usable local cursor (i.e., no cursor at all, or an all-transparent cursor), too.
